cachi2/core/package_managers/generic/main.py

```diff
@@ -47,8 +53,47 @@ def _resolve_generic_lockfile(source_dir: RootedPath, output_dir: RootedPath) ->
 
     log.info(f"Reading generic lockfile: {lockfile_path}")
     lockfile = _load_lockfile(lockfile_path)
-    for artifact in lockfile.artifacts:
+    artifacts: List[GenericArtifact] = [
+        GenericArtifact.from_lockfile_artifact(a) for a in lockfile.artifacts
+    ]
+
+    # output_dir is now the root and cannot be escaped
+    output_dir = output_dir.re_root(DEFAULT_DEPS_DIR)
+
+    for artifact in artifacts:
         log.debug(f"Resolving artifact: {artifact.download_url}")
+        # make sure target does not point outside the output directory
+        try:
+            destination = output_dir.join_within_root(artifact.resolved_target)
+        except PathOutsideRoot as e:
+            e.solution = (
+                f"Artifact '{artifact.download_url}' has target which falls "
+                "outside of the dependencies root. Refusing to proceed."
+            )
+            raise
+
+        # make sure that targets do not overlap
+        if any([a.has_same_resolved_path(destination) for a in artifacts]):
+            raise PackageRejected(
+                f"More than one file from {lockfile_path} is trying to write to the same target: {destination}",
+                solution="Make sure that all targets are unique.",
+            )
+
+        # create the parent directory and assign the resolved path to the artifact
+        Path.mkdir(destination.path.parent, parents=True, exist_ok=True)
+        artifact.resolved_path = destination
```

Comment:
This does not make sense to me. There is an attribute in the class that is Optional and is not set when creating the object. It is only overwritten here. In addition, you try to do a comparison via `has_same_resolved_path` even if it is None.

Reply:
I feel like the `GenericArtifact` class is not bringing too much value, it seems mostly redundant. Maybe we can add the `resolved_path` attribute in the `LockfileArtifact` class itself, or even rely on `pathlib.Path(GenericArtifact.target).resolve()` to look for conflicts and drop this attribute altogether.

Reply:
Removed the `GenericArtifact` class, it was mostly redundant. I originally created it, because I did not want to mix validation and whatever happens next to the artifact. I've added  `_resolved_path` in the `LockfileArtifact` class as a private attribute, so it couldn't be loaded from the lockfile, but could still be used to store the path. Thanks for the feedback, I've definitely overengineered it.